lib/usd/translators/shading/rfmShaderTranslation.cpp

```diff
@@ -0,0 +1,62 @@
+//
+// Copyright 2020 Pixar
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "genericShaderWriter.h"
+
+#include <mayaUsd/fileio/shaderWriterRegistry.h>
+#include <mayaUsd/fileio/shading/rfmShaderMap.h>
+#include <mayaUsd/fileio/shading/shadingModeRegistry.h>
+
+#include <pxr/pxr.h>
+#include <pxr/base/tf/registryManager.h>
+#include <pxr/base/tf/staticTokens.h>
+#include <pxr/base/tf/token.h>
+
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+
+TF_DEFINE_PRIVATE_TOKENS(
+    _tokens,
+
+    ((conversionName, "rendermanForMaya"))
```

Comment:
@mattyjams Would it be problematic if we make this translator a separate plugin?

Reply:
@kxl-adsk: I had thought of that, but I was a little reluctant to do so right now. My short term goal here is to get rid of the `pxrRis` shadingMode and replace it with this. Anyone who had been using that would need to start using the `useRegistry` shadingMode instead, so I was hesitant to force them to make that change *and* load something else. Especially so since with the use of the genericShaderWriter, there would be almost nothing to the plugin. If you all have strong feelings that this be separated out though, I could do that. I'd just need some guidance as to what structure/naming would make sense for that.

I think ideally, this could one day be owned by the RenderMan team and just ship as part of their plugin once this shading-related API stabilizes and is available in public releases.

Reply:
I see your argument for not having much left in the plugin. Moving it over to RenderMan team would match our own discussions around Arnold and translators for Arnold.

How about we then make this translator conditionally compiled? The point to make here is that not everyone has access to RenderMan and we are considering taking it out of the binaries we release.

Reply:
Yeah, that makes sense. I can see not necessarily wanting to ship RenderMan-related stuff with the base mayaUsd installer, though there isn't currently a way to cull the `pxrRis` shadingMode. But hopefully we're killing that soon anyway!

In the rebased commits I just pushed, you'll see a `BUILD_RFM_TRANSLATORS` build option was added in https://github.com/Autodesk/maya-usd/pull/787/commits/fc8efb0b686310a2e87c46fe514d5d83aeff979f. This needed to go at the top-level since the option is used in both the `lib/...` and `test/...` trees, and you'll see the option checked again to conditionally add the test in  https://github.com/Autodesk/maya-usd/pull/787/commits/2d951e4b008db3146b353ee1403630cc0512b9b0.

Reply:
Thank you for adding `BUILD_RFM_TRANSLATORS`, @mattyjams.

I'm definitely for killing pxrRis shadingMode - is this something you are planning on doing in the near future or it's more on us to execute on it?

Reply:
Yeah, I'm planning on doing that once `useRegistry` provides a complete replacement for the functionality `pxrRis` provided. We don't actually use the `pxrRis` shadingMode internally, so most of the work is probably just making sure that anyone external who does use it is aware that it's being removed.

I'm planning to kill the `displayColor` shadingMode too, though that might be slightly more contentious since it seems more likely that people actively use that one. Hopefully not though. With Jerry's work to support Maya-native materials, I don't see much of a need to keep `displayColor` anymore.

Reply:
> I'm planning on doing that once useRegistry provides a complete replacement for the functionality pxrRis provided

What would you consider as still missing?

Reply:
This PR only adds export support, so I have a follow-up one I'm working on that'll take care of import.

A few nuances off the top of my head that we'll still need to deal with:
- exporting needs to be able to consider non-built-in shading plugs that some renderers use on the `shadingEngine` node:
https://github.com/Autodesk/maya-usd/blob/91d0bfe820f74e65d4d4f91049d8171fbf1b5dcf/lib/mayaUsd/fileio/shading/shadingModePxrRis.cpp#L83
RenderMan for Maya obviously does this, but I think Arnold might as well? I think the intent is to help differentiate between the shader used for previewing in the Maya viewport and the renderer-specific shader used by the renderer.
- Similarly, import likely needs to support creating these non-built-in plugs and making shading connections to them, for example when a Material prim has both an `outputs:surface` and an `outputs:ri:surface` terminal, or potentially if we want to start considering [materialPurpose (e.g. preview vs. full)](https://github.com/PixarAnimationStudios/USD/blob/release/pxr/usd/usdShade/materialBindingAPI.h#L77).
- `useRegistry` currently only supports importing the universal renderContext terminals of Material prims. The export of RfM shaders here writes them into the `ri` renderContext, so `useRegistry` will not currently see them on import. This will tie into the feature above so that the import knows which plugs to connect these shaders to on the `shadingEngine`.

I'm working on (or will be working on) all of these things and will put up PRs for them, and happy to discuss the intricacies of each as we dig into them.

Reply:
Seems we need to get in touch as I am also looking into expanding the material import code to support an import path similar to the materialConversion that was added recently on export.
Some information:
- Arnold does use the same shading group attachment points as regular surface materials, so it does not allow a material purpose concept in Maya.
- We would like to be able to pass multiple shadingMode to the import command to allow specifying a list of locations where we want to search for materials in USD. I thought the search would end on the first surface material found, but having custom entry points for RfM might mean reconsidering how the search terminates. The import options would be simplified to try everywhere, and a user wanting full control over where to search could do so at the command level.
- The usdListShadingModes command would be expanded to also list materialConversions that have an import path. This means expanding the materialConversion registry with import information.

Reply:
@JGamache-autodesk: Yup, definitely. Sorry, I didn't mean to suggest that any of the things I mentioned were solved problems with well understood solutions that just needed implementing. I'm sure we'll be talking about all of these things and more as we keep going. PRs (or maybe draft PRs?) might be a good way to anchor some of the discussion around something concrete as we break down each piece, but I'm expecting we'll need to iterate a bit to find the best way forward.

Good to know about Arnold, as well as the other ideas you have in mind. All sounds great to me. Catch up with you about it soon!